apps/org/views.py

```diff
@@ -461,76 +480,117 @@ class OrgCreateMemberCompleteView(LoginRequiredMixin, OrgCreateMemberMixin, Form
+        return token_generator.check_token(user, token)
+
+    def get(self, *args, **kwargs):
+        """GETting the OrgCreateMemberCompleteView is only allowed with a valid token."""
+        self.organization = self.get_organization(self.request, self.kwargs.get('org_slug'))
+        self.member = self.get_member(self.organization, self.kwargs.get('username'))
+
+        if self.token_is_valid(uidb64=kwargs['uidb64'], token=kwargs['token']):
+            # The token is valid, so render the page to the user
+            return self.render_to_response(self.get_context_data())
+        else:
+            # The token is not valid, so redirect user to the org_create_member_invalid_token page
+            redirect_url = reverse(
+                'org:org_create_member_invalid_token',
+                kwargs={'org_slug': kwargs['org_slug'], 'username': kwargs['username']}
+
+            )
+            return HttpResponseRedirect(redirect_url)
+
+    def post(self, *args, **kwargs):
+        """POSTing to the OrgCreateMemberCompleteView is only allowed with a valid token."""
+        if self.token_is_valid(uidb64=kwargs['uidb64'], token=kwargs['token']):
+            # The token is valid, so call the super().post() method
+            return super().post(*args, **kwargs)
+        else:
+            # The token is not valid, so redirect user to the org_create_member_invalid_token page
+            redirect_url = reverse(
+                'org:org_create_member_invalid_token',
+                kwargs={'org_slug': kwargs['org_slug'], 'username': kwargs['username']}
+
+            )
+            return HttpResponseRedirect(redirect_url)
+
     def form_valid(self, form):
         """
         If form is valid, then create a ResourceGrant, set password, redirect user to the next step.
 
         If the form is valid, that means that the member is approving the Organization's
         ResourceRequest, and is also setting their password. Therefore, we
-         1.) find and update the relevant ResourceRequest to be approved
-         2.) create a ResourceGrant object
-         3.) verify that the request.user has a UserSocialAuth object for VMI
-         4.) verify that the Member has a UserSocialAuth object for VMI
-         5.) make a request to VMI to update the member's password in VMI
+         1.) verify that the request.user (the Member) has a UserSocialAuth object for VMI
+         2.) find and update the relevant ResourceRequest to be approved
+         3.) create a ResourceGrant object
+         4.) make a request to VMI to update the member's password in VMI
+         5.) set the member's password locally
          6.) redirect the user to the next step in the Member-creation process
         """
-        # 1.) Find and update the ResourceRequest for this Member to be approved
+        # 1.) Verify that the request.user has a UserSocialAuth object for VMI
+        request_user_social_auth = self.request.user.social_auth.filter(
```

Comment:
results in an error if the link is followed with no session because `self.request.user` is the anonymous user.

Should be fixable by logging the user associated with the unique token in.

Reply:
May be able to handle this nicely the `LoginRequired` mixin, or a custom version of it, and some custom authentication backend work that handles checking the token.

Reply:
Good 👁. I've updated the code, though I didn't want to log the user in, since it wasn't clear to me that they would know they are logged in after this step (especially if they're on the organization's device)